osf/db/router.py

```diff
@@ -0,0 +1,70 @@
+from django.conf import settings
+import psycopg2
+
+CACHED_MASTER = None
+
+
+class PostgreSQLFailoverRouter(object):
+    """
+    1. CHECK MASTER_SERVER_DSN @ THREAD LOCAL
+    2. THERE?, GOTO 9
+    3. GET RANDOM_SERVER FROM `settings.DATABASES`
+    4. CONNECT TO RANDOM_SERVER
+    5. IS MASTER SERVER?
+    6. YES? GOTO 8
+    7. NO?, `exit()`
+    8. STOR MASTER_SERVER_DSN @ THREAD_LOCAL
+    9. PROFIT
+    Number of servers can be assumed to be > 1 but shouldn't assume 2 max.
+    Might be nice to keep track of the servers that have been tried from settings.DATABASES so we don't get into a loop.
+    """
+    DSNS = dict()
+
+    def __init__(self):
+        self._get_dsns()
+        global CACHED_MASTER
+        if not CACHED_MASTER:
+            CACHED_MASTER = self._get_master()
+
+    def _get_master(self):
+        for name, dsn in self.DSNS.iteritems():
+            conn = self._get_conn(dsn)
+            cur = conn.cursor()
+            cur.execute('SELECT pg_is_in_recovery();')
```

Comment:
This is good but lets stick with the JDBC implementation if at all possible.

https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/core/v3/ConnectionFactoryImpl.java#L692